### PR TITLE
Adding sync pod latency metric.

### PR DESCRIPTION
**Pull request by vmarmol**
Latency is broken down by create, update, and sync pods.

Part of #4604.

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1336,6 +1339,15 @@ func (kl *Kubelet) SyncPods(pods []api.BoundPod) error {
...
 			}
+
+			// Update metrics for this pod operation.
+			if _, ok := newPods[pod.UID]; ok {
+				metrics.SyncPodLatency.WithLabelValues("create").Observe(metrics.SinceInMicroseconds(start))
+			} else if _, ok := updatedPods[pod.UID]; ok {
+				metrics.SyncPodLatency.WithLabelValues("udpate").Observe(metrics.SinceInMicroseconds(start))
+			} else {
+				metrics.SyncPodLatency.WithLabelValues("sync").Observe(metrics.SinceInMicroseconds(start))
+			}
```
Comment (rjnagal) [inline pkg/kubelet/kubelet.go]: Collapse these three calls to one using  a map from podid to operation type.

File: pkg/kubelet/metrics/metrics.go [hunk 2]
```diff
@@ -30,30 +32,26 @@ var (
...
-			Subsystem: kubeletSubsystem,
-			Name:      "pod_count",
-			Help:      "Number of pods currently running.",
-		},
-	)
-	// TODO(vmarmol): Implement.
-	// TODO(vmarmol): Split by source?
-	ContainerCount = prometheus.NewGauge(
-		prometheus.GaugeOpts{
+	// TODO(vmarmol): Break down by number of containers in pod?
```
Comment (rjnagal) [inline pkg/kubelet/metrics/metrics.go]: +1

File: pkg/kubelet/metrics/metrics.go [hunk 3]
```diff
@@ -30,30 +32,26 @@ var (
...
-			Name:      "container_count",
-			Help:      "Number of containers currently running.",
+			Name:      "sync_pod_latency_microseconds",
+			Help:      "Latency in microseconds to sync a single pod. Broken down by operation type: create, update, or sync",
 		},
+		[]string{"operation_type"},
 	)
 	// TODO(vmarmol): Containers per pod
-	// TODO(vmarmol): Latency of pod startup
 	// TODO(vmarmol): Latency of SyncPods
```
Comment (rjnagal) [inline pkg/kubelet/metrics/metrics.go]: what else do we need to do for SyncPods?

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Good call, done.

File: pkg/kubelet/metrics/metrics.go (hunk 3 above)
Comment (vmarmol) [inline pkg/kubelet/metrics/metrics.go]: Nothing much, it should just be a new definition and defer call. Will probably be in a followup PR :)

**Comment by vmarmol** [comment]
Done @rjnagal PTAL :)

File: pkg/kubelet/metrics/metrics.go [hunk 4]
```diff
@@ -47,10 +56,37 @@ func Register(containerCache dockertools.DockerCache) {
...
+)
+
+func (self SyncPodType) String() string {
+	switch self {
+	case SyncPodCreate:
+		return "create"
+	case SyncPodUpdate:
+		return "create"
+	case SyncPodSync:
+		return "create"
```
Comment (rjnagal) [inline pkg/kubelet/metrics/metrics.go]: Why is everything "create" ? :)

Comment (vmarmol) [inline pkg/kubelet/metrics/metrics.go]: lol, because I suck at copy and paste :) I got unknown right though :D

Fixed.

**Comment by vmarmol** [comment]
Argh, the recent merge made this harder. Will fix and let you know when it is safe once more.

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -61,7 +61,10 @@ const podOomScoreAdj = -100
 
 // SyncHandler is an interface implemented by Kubelet, for testability
 type SyncHandler interface {
-	SyncPods([]api.BoundPod) error
+	// Syncs current state to match the specified pods. SyncPodType specified what
+	// type of sync is occuring per pod. StartTime specifies the time at which
+	// synching began (for use in monitoring).
```
Comment (rjnagal) [inline pkg/kubelet/kubelet.go]: nit: s/synching/syncing/ or not? idk.

**Comment by rjnagal** [comment]
Also needs a rebase.

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Done.

**Comment by vmarmol** [comment]
Fixed and rebased PTAL @rjnagal

File: pkg/kubelet/pod_workers.go [hunk 6]
```diff
@@ -64,29 +72,33 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan api.BoundPod) {
...
 		// TODO(wojtek-t): Adjust the size of the buffer in this channel
-		podUpdates = make(chan api.BoundPod, 5)
+		podUpdates = make(chan workUpdate, 5)
 		p.podUpdates[uid] = podUpdates
 		go p.managePodLoop(podUpdates)
 	}
-	podUpdates <- pod
+	podUpdates <- workUpdate{
+		pod:               pod,
+		updateCompleteFun: updateComplete,
```
Comment (rjnagal) [inline pkg/kubelet/pod_workers.go]: heh, Fun.

**Comment by rjnagal** [comment]
LGTM. waiting for travis.

File: pkg/kubelet/pod_workers.go (hunk 6 above)
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: lol, This updates the Complete Fun (TM)

Can change to Func if you think it's better (I tend to agree). I was keeping the style of the file.

Comment (rjnagal) [inline pkg/kubelet/pod_workers.go]: We have Fn, Fun, and Func. I wouldn't say no to Fun.
You can refactor in a separate PR if you want. Let's get this one in.

**Comment by ixdy** [comment]
FYI, this was reverted by #4867 because it broke e2e tests on gce.

**Pull request merged**